File: dictionnary/dict.cpp
#include <iostream>
#include <fstream>
#include <map>
#include <string>

struct Node {

    void record(
            std::string_view::const_iterator first, 
            std::string_view::const_iterator last)
    {
        if (first == last) {
            is_terminal = true;
            return;
        }
        Node &next = children[*first];
        next.record(first+1, last);
    }

    bool is_terminal = false;
    std::map<char, Node> children;
};

class Dict {
    
public:

    void record(std::string_view str)
    {
        root.record(str.begin(), str.end());
    }

    void print()
    {
        std::cout << "Dictionnary content:\n";
        print_node(root, "");
        std::cout.flush();
    }

private:

    static 
    void print_node(Node const &node, std::string str)
    {
        if (node.is_terminal) {
            std::cout << "  --> " << str << '\n';
        }
        for (const auto &[ch, next] : node.children) {
            print_node(next, str + ch);
        }
    }

private:
    Node root;
};

int main() {

    Dict dict;
    std::ifstream in("dict.txt");
    if (!in) {
        std::cerr << "File \"dict.txt\" is invalid!" << std::endl;
        return 1;
    }
    std::string buffer;
    while (!in.eof()) {
        in >> buffer;
        dict.record(buffer);
    }
    //dict.print();
    return 0;
}

File: dictionnary/dict.rs
pub mod dict {

    use std::collections::BTreeMap;
    use std::str::Chars;

    #[derive(Default, Clone)]
    pub struct Node {
        pub is_terminal: bool,
        children: BTreeMap<char, Node>
    }

    #[derive(Default, Clone)]
    pub struct Dict {
        pub root: Node
    }

    impl Node {
        pub fn record(&mut self, word: &mut Chars)
        {
            let c: char;
            match word.next() {
                Some(ch) => c = ch,
                None => {
                    self.is_terminal = true;
                    return;
                }
            };
            let node = self.children.entry(c).or_insert(Node::default());
            node.record(word);
        }
    }

    impl Dict {
        pub fn record(&mut self, str: &str)
        {
            self.root.record(&mut str.chars());
        }

        fn print_node(node: &Node, str: String)
        {
            if node.is_terminal {
                println!(" --> {}", &str);
            }
            for (ch, next) in &node.children {
                let mut str_cpy = str.clone();
                str_cpy.push(*ch);
                Dict::print_node(next, str_cpy);
            }
        }

        pub fn print(&self)
        {
            println!("Dictionnary content:");
            Dict::print_node(&self.root, "".to_string());
        }
    }
}

fn main() {
    let mut dict = dict::Dict::default();
    dict.record("Hello");
    dict.record("Hulk");
    dict.record("Hat-trick");
    dict.record("World");
    dict.record("Wonderful");
    dict.record("Rust");
    dict.print();
}
